### PR TITLE
change markdown parser to mistletoe

### PR DESCRIPTION
**Pull request by ghost**
Switched Markdown parser to [Mistletoe](https://github.com/mataroa-blog/mataroa/issues/33).

Related to [this](https://github.com/mataroa-blog/mataroa/issues/33) issue.

### CODE REVIEW

File: requirements.in [hunk 1]
```diff
@@ -5,3 +5,4 @@ markdown
 pygments
 bleach[css]
 stripe
+mistletoe
```
Comment (sirodoht) [inline requirements.in]: Could we add a new line here at the end of this file. We have an undocumented convention to do this, note to self, to document it!

File: main/util.py [hunk 2]
```diff
@@ -4,13 +4,16 @@
 import zipfile
 
 import bleach
-import markdown
+import mistletoe
+from mistletoe import HTMLRenderer
```
Comment (sirodoht) [inline main/util.py]: Would it be possible to remove this line and access HTMLRenderer via `mistletoe.HTMLRenderer` below?

**Review by sirodoht — CHANGES_REQUESTED** [review]
Thanks for submitting this! It's great, just a couple of preferences changes

**Comment by sirodoht** [comment]
As far as the test failing, it seems mistletoe adds a new line by default to any single line renderings. To fix the test failing, I think you can change the last line of the `main/tests/test_feeds.py` file from:

```py
f"<description>{self.user.blog_byline}</description>"
```

to:

```py
f"<description>{self.user.blog_byline}\n</description>"
```

The change is an addition of a `\n` character in the content of the tag.

**Comment by ghost** [comment]
No problem! I'll do that when I get home, in a few hours.

**Comment by sbstnlol** [comment]
Hey! I've been working on the changes, and I decided to run ``python3 manage.py test`` locally, and I get the following error:

```
Traceback (most recent call last):
  File "/home/sebastian/Desktop/projects/web/mataroa/main/tests/test_images.py", line 103, in test_image_raw
    self.assertEqual(self.image.data.tobytes(), response.content)
                     ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'bytes' object has no attribute 'tobytes'
```

Do you know what's wrong? Nothing I modified has to do with images at all.

File: main/tests/test_images.py [hunk 3]
```diff
@@ -100,7 +100,7 @@ def test_image_raw(self):
             reverse("image_raw", args=(self.image.slug, self.image.extension)),
         )
         self.assertEqual(response.status_code, 200)
-        self.assertEqual(self.image.data.tobytes(), response.content)
```
Comment (sirodoht) [inline main/tests/test_images.py]: That's probably due to how django saves binary data on sqlite versus postgres

**Comment by sirodoht** [comment]
I think I've made it so that you don't need my approval to run the CI anymore

**Comment by ghost** [comment]
All checks have passed now!

**Comment by sirodoht** [comment]
Great! I can’t recall if we talked about this but are the extensions we removed now supported natively? Namely tables, footnotes, toc.

**Comment by ghost** [comment]
> Great! I can’t recall if we talked about this but are the extensions we removed now supported natively? Namely tables, footnotes, toc.

I don't know what ``toc`` is, but give me a few minutes and I'll test it!

**Edit: It's "table of contents". Testing now...**

**Comment by ghost** [comment]
Tables work, ToC doesn't, and neither do footnotes. Working on both now.

**Comment by sbstnlol** [comment]
@sirodoht Can you give me a proper example of Markdown ToC? I want to make sure I'm testing the right thing.

**Comment by ghost** [comment]
I want to get this working, but I don't know if I can figure it out. There isn't very detailed documentation for mistletoe, so it isn't the easiest thing to work with.

**Comment by sirodoht** [comment]
In terms of the TOC, I made an example here: https://nutcroft.com/blog/a-new-startup-lifecycle/

Essentially, the important part is that headings have ids. For example the first heading is an h2 and its id is `i-the-philosophy-of-posthaven`.

In other words it renders as `<h2 id="i-the-philosophy-of-posthaven">I. The philosophy of Posthaven</h2>`

For this reason, we can have a link such as `https://nutcroft.com/blog/a-new-startup-lifecycle/#i-the-philosophy-of-posthaven` which will scroll to that heading in the article.

The least important part, which we can do without, is the `[TOC]` marker, which generates these links in the beginning:
image

Let me know if these make sense!

**Comment by sirodoht** [comment]
In terms of footnotes, I see that mistletoe does not support them out of the box: https://github.com/miyuchina/mistletoe/issues/47

One question is how easy we can re-implement them by overriding their HTMLRenderer, just like we did with the code highlighting part.

They don't have a footnotes example in their contrib folder as they do for code highlighting: https://github.com/miyuchina/mistletoe/tree/e853a0def73f4d4c2be5bf16ef4ef4135df0d556/mistletoe/contrib

But there is some kind of reference to footnotes in their docs building page here:
https://github.com/miyuchina/mistletoe/blob/e853a0def73f4d4c2be5bf16ef4ef4135df0d556/docs/__init__.py#L26-L34

as well as a `Footnote` BlockToken class, though I haven't figured out yet if that's the footnote we mean:
https://github.com/miyuchina/mistletoe/blob/e853a0def73f4d4c2be5bf16ef4ef4135df0d556/mistletoe/block_token.py#L745

**Pull request closed without merging**